### ISSUE TITLE
Optimize canvas renderer

### ISSUE THREAD

**Issue by bartvde** [openlayers/openlayers#4328]
This is a follow-up ticket from #4219 
excerpt from @ahocevar:
"I spent a long time profiling and optimizing. The bottleneck is now in the renderer code itself. The majority of the time is spent applying strokes and fills. If the canvas renderer did a better job combining geometries and stroking/filling them at once, performance would be better. But that's a separate issue, as it affects all canvas rendering in OpenLayers. The goal is to reduce the number of stroke and fill instructions."

**Comment by ahocevar** [comment]
A quick investigation (done on Chrome 48) shows that there are at least two things that can be improved. See the following change:
```diff
--- a/src/ol/render/canvas/canvasreplay.js
+++ b/src/ol/render/canvas/canvasreplay.js
@@ -460,6 +460,8 @@ ol.render.canvas.Replay.prototype.replay_ = function(
         ++i;
         break;
       case ol.render.canvas.Instruction.SET_STROKE_STYLE:
+      this.prevStrokeStyle_ = undefined;
+      this.prevLineWidth_ = undefined;
         goog.asserts.assert(goog.isString(instruction[1]),
             '2nd instruction should be a string');
         goog.asserts.assert(goog.isNumber(instruction[2]),
@@ -500,7 +502,12 @@ ol.render.canvas.Replay.prototype.replay_ = function(
         ++i;
         break;
       case ol.render.canvas.Instruction.STROKE:
-        context.stroke();
+        if (this.prevStrokeStyle_ !== context.strokeStyle || this.prevLineWidth_ !== context.lineWidth) {
+          context.stroke();
+          console.log(context.strokeStyle, context.lineWidth)
+          this.prevStrokeStyle_ = context.strokeStyle;
+          this.prevLineWidth_ = context.lineWidth;
+        }
         ++i;
         break;
       default:
```

1. `context.stroke()` (and probably also `context.fill()`) are currently called even when the stroke style has not changed. When removing the `style` configuration for the mapbox-vector-tiles example, the above stroke instruction is called 16 times. Without the change, it is called 174 times.
2. Grouping features by the same style could improve things even more. With the unaltered mapbox-vector-tiles example, the stroke style often changes, but the total number of different stroke styles is quite limited.

According to https://jsperf.com/batch-canvas-stroke-fill-calls, applying strokes to batches is twice as fast as applying them to individual geometries. In the tests above, I could only notice a slight improvement by batching the stroke instructions, but that's probably because the strokeStyle was still set more often than necessary. With my change, less than 50% of the time was spent applying strokes. Without my change, it was more than 50%.

So there clearly is room for improvement, but it won't be orders of magnitude.

**Comment by ahocevar** [comment]
Also see #4393.

**Comment by fredj** [comment]
Fixed with #5149 ?

**Comment by ahocevar** [comment]
Fixed with #5149.

**Comment by jeroenvheel** [comment]
The code above isn't working properly the outer line on a polygon is isn't drawn.
